### ISSUE TITLE
Domain Security Issue

### ISSUE THREAD

**Issue by SjamonDaal** [serghey-rodin/vesta#1524]
### Operating System (OS/VERSION):
CentOS 7

### VestaCP Version:
0.9.8-19

### Installed Software (what you got with the installer):
php-fpm, apache, nginx, named

### Other Notes:

Issue example.
User 1 owns the domain example.com, he creates a website and the hoster is using a VestaCP DNS cluster.
user 1 creates the website with exaple.com and www.example.com.

User 2 is also registered at the hoster's VestaCP but doen't own a domain name. he found example.com hosted on the same system. User 2 registers xxx.example.com (a subdomain).

User 2 is now using the domain owned by User 1.

This is a security issue. Please lock domains per user. If User 1 owns *.example.com, then ok else if user 2 owns *.example.com, then deny access. "This domain is registered to another user."

**Comment by SjamonDaal** [comment]
After the fix of 0.9.8-20,
I found some code I could re-use in the commit,
https://github.com/serghey-rodin/vesta/commit/eaf9d89096b11daa97f8da507eb369e359cda7dd

Building a function on site creation tonight.

**Comment by lazzurs** [comment]
This is an ownership issue but in some cases shared ownership will be desirable.

A hosting provider might want to allow say $customer.awesomewebsites.com where awesomewebsites.com is under the hosting company user and the $customer.awesomewebsites.com are under the account of the users. There maybe other use cases of shared domains.

It would be good to have a flag on a domain to indicate the domain is shared, basically the current behaviour, or single user in which case only one user gets to create DNS entries.

**Comment by SjamonDaal** [comment]
@lazzurs is think this is a **Security Issue** and it needs to get some attention quickly.

If a domain is registered and the DNS settings are pointing to your vestacp DNS Cluster, it's not only possible to create a subdomain and DNS records. It's also possible to create a mailbox on the subdomain.

Say I own for example github.com and I register it with my hosting provider. We mail from our own mail servers for example 0365, we user @github.com as mail domain.

Some other random guys is able to create a subdomain for example support.github.com, DNS records for the VestaCP mail server are automatically added. Now i the random guy just adds a mail domain like support.guthub.com and creates a mailboxes like jimmy@support.github.com.

Looks legit for all receivers.

This is a Security Issue.

**Comment by lazzurs** [comment]
@SjamonDaal I agree there is an issue here and I understand it well, I was not disputing that. I was just making clear that this behaviour may be desirable for some domains and may be the intent.

I would suggest it should be optional to stop this behaviour as it may suit some users as it is.

**Comment by SjamonDaal** [comment]
@lazzurs Agree! But I do think this should be disabled by default and only optionally enabled from the admin user / api for the desired domain name.

**Comment by Neustradamus** [comment]
Any solutions?
It is resolved?

**Comment by SjamonDaal** [comment]
@Neustradamus, No this still needs to be resolved.

**Comment by anton-reutov** [comment]
We don't think that is a problem. This behavior has been always

**Comment by insound** [comment]
I agree about vestacp should have a checkbox in install script or in settings in web (better) to check main domain owners and don't create wildcard or subdomain when another user trying to do that.
I guess its small fix to check users dns config in vestacp file dns before to process bin file, wich use $ in settings vestacp. U can do it by yourself, because vestacp always been with this "issue".

**Comment by SjamonDaal** [comment]
@anton-reutov

> We don't think that is a problem. This behavior has been always

So it's a bug by design. But I brought this to the attention because of a security issue that's going on here.

I think this system should be secure against fraud protection.

**Comment by ifaist0s** [comment]
> 
> 
> We don't think that is a problem. This behavior has been always

This "behavior" can have other serious implications. Consider the following hypothetical scenarios:
### Scenario 1
Let's assume that @anton-reutov has registered the domain vestacp.com with his real name. Let's also assume that the domain is hosted on a shared VESTA environment.

Other malicious entities, who are aware of this "behavior", decide to take advantage of it and host their illicit web site / service on the same VESTA environment, creating the subdomain **silkroad.vestacp.com**.

After some time, the FBI learns about the illicit web site silkroad.vestacp.com and decides to take action against it. Who's gonna be arrested? @anton-reutov of course! The owner of the domain name.

**If you think this "behavior" is not a problem, stand up for your opinion and let us host our sites/services on the same VESTA installation that vestacp.com is hosted on.**

### Scenario 2
Let's assume that @anton-reutov owns a small company, providing hosting services to customers. Let's also assume that he's using VESTA CP for his business.

One (or more) of his clients, fall victim of this "behavior" because other malicious clients exploited it like in the above scenario. Legit clients are dragged to court to prove they are innocent and in turn they take legal action against @anton-reutov to claim their damages.

### Conclusion
Maybe you should re-think the whole thing. Unless you think that VESTA CP is for personal use only. In that case, I would agree with you that this "behavior" is not a problem. In any other case, what @lazzurs suggested seems to be the middle ground, especially if it's provided as Opt-In (disabled by default - enable when desired).
